### PR TITLE
Remodel input, adjust components which uses it

### PR DESCRIPTION
**Pull request by AlexeyKaryka**
## This PR is a:

[ ] New feature
[ ] Enhancement/Optimization
[ ] Refactor
[ ] Bugfix
[ ] Test for existing code
[ ] Documentation

Update Input component, adjust createAccount, signIn

## Summary

When this pull request is merged, it will...

### CODE REVIEW

File: packages/venia-concept/src/components/Input/helpers.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    for (let i = 0; i < validators.length; i++) {
+        const validatorOutput = validators[i](value, values);
+
+        if (validatorOutput) {
+            return validatorOutput;
+        } else if (i + 1 < validators.length) {
+            // If current iteration is on the last validator
+            // and this validator hasn't returned error
+            // we need to return null as result of whole complex valdiator
+            continue;
```
Comment (Starotitorov) [inline packages/venia-concept/src/components/Input/helpers.js]: You do not need to do this. It is better just to return null after this for cycle.

File: packages/venia-concept/src/components/Input/helpers.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { isString } from 'util';
+
+const emptyFieldMessage = 'This field cannot be empty';
+export const defaultIsRequired = value =>
```
Comment (Starotitorov) [inline packages/venia-concept/src/components/Input/helpers.js]: Could you just change the name of the function to isRequired?

File: packages/venia-concept/src/components/Input/helpers.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    isString(value) && value.length > 0 ? null : emptyFieldMessage;
+
+export const createComplexValidator = validators => (value, values = null) => {
```
Comment (Starotitorov) [inline packages/venia-concept/src/components/Input/helpers.js]: I believe in general it is better to at least move this to util/ folder. Or create services/ folder and create validator.js file there.

File: packages/venia-concept/src/components/Input/input.js [hunk 4]
```diff
@@ -32,48 +33,76 @@ class Input extends Component {
         required: PropTypes.bool,
         title: PropTypes.string,
         autoComplete: PropTypes.string,
-        helpText: PropTypes.string,
         helpType: PropTypes.string,
+        helpText: PropTypes.string,
         field: PropTypes.string.isRequired,
-        onChange: PropTypes.func
+        onChange: PropTypes.func,
+        validate: PropTypes.func,
```
Comment (Starotitorov) [inline packages/venia-concept/src/components/Input/input.js]: You do not need this prop types definition, asField hoc manages this

File: packages/venia-concept/src/components/SignIn/signIn.js [hunk 5]
```diff
@@ -1,12 +1,18 @@
 import React, { Component } from 'react';
 import PropTypes from 'prop-types';
+import get from 'lodash/get';
```
Comment (Starotitorov) [inline packages/venia-concept/src/components/SignIn/signIn.js]: Right now we did not define lodash as a dependency in our package.json, but I think that `get` function from lodash is extremely useful. @zetlen, @jimbo Could we discuss adding lodash.get https://github.com/lodash/lodash/tree/4.4.2-npm-packages/lodash.get to the project?

File: packages/venia-concept/src/components/SignIn/signIn.js [hunk 6]
```diff
@@ -17,44 +23,48 @@ class SignIn extends Component {
...
         signIn: PropTypes.func
     };
 
-    state = {
-        password: '',
-        username: ''
-    };
+    constructor(props) {
+        super(props);
+        this.formState = {};
```
Comment (Starotitorov) [inline packages/venia-concept/src/components/SignIn/signIn.js]: In general, it is a bad idea to put the data to this object directly. In this case component will not rerender when something changes

Comment (Starotitorov) [reply packages/venia-concept/src/components/SignIn/signIn.js]: In general, you do need to store something in local state. Form component from informed can retrieve function as a children. So, formState will be passed to this function as a property.

File: packages/venia-concept/src/components/SignIn/signIn.js [hunk 7]
```diff
@@ -76,22 +86,15 @@ class SignIn extends Component {
     }
 
     onSignIn = () => {
-        const { username, password } = this.state;
-        this.props.signIn({ username: username, password: password });
+        const username = this.getFieldValue(fields.username);
```
Comment (Starotitorov) [inline packages/venia-concept/src/components/SignIn/signIn.js]: If you pass onSignIn function to onSubmit will receive `values` object as an argument.

File: packages/venia-concept/src/components/Checkout/formContainer.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+import classify from 'src/classify';
+import Form from './form';
+import defaultClasses from './form.css';
+
+const mapStateToProps = ({
+    checkout: { isAddressIncorrect, incorrectAddressMessage }
+}) => ({ isAddressIncorrect, incorrectAddressMessage });
+
+export default compose(
+    classify(defaultClasses),
```
[the remark this answers is not in the document]
Comment (AlexeyKaryka) [reply packages/venia-concept/src/components/Checkout/formContainer.js]: @jimbo I've seen that in all components with containers classify wraps common component in container's export, not in common component's export, could you please describe why this approach has been chosen?

File: packages/venia-concept/src/components/Input/helpers.js (hunk 3 above)
Comment (AlexeyKaryka) [reply packages/venia-concept/src/components/Input/helpers.js]: https://github.com/magento-research/pwa-studio/pull/593#discussion_r238682240

**Pull request closed without merging**